flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java

```diff
@@ -427,13 +427,7 @@ public void validateRunsInMainThread() {
 
         @Override
         public void execute(@Nonnull Runnable command) {
-            if (!scheduledExecutorService.isShutdown()) {
```

Comment:
We can not ignore even one runnable here since MainThreadExecutor can be the executor for CompleatableFuture and if scheduledExecutorService would be shutdown before all futures will be done we can lost important execution like releasing resources. For example, take a look at org.apache.flink.runtime.scheduler.SchedulerBase#closeAsync where shutDownCheckpointServices is called exactly as I described before.

Reply:
You are right. @zjureel I think we need to do the same investigation for `schedule` as well.